feat: replace std Instant with wasm-compatible wrapper

waynexia opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Relates to https://github.com/apache/arrow-datafusion/issues/7651 and #7652

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

`std::time::Instant` doesn't work on target `wasm32-unknown-unknown`. It would result in a runtime panic when using it. A detailed explanation can be found in the replacement wrapper crate `instant`'s [repository](https://github.com/sebcrozet/instant?tab=readme-ov-file#instant).

It doesn't have a difference when runs on other platforms. 

## What changes are included in this PR?

Replace all the occurrences of `std::time::Instant` with `instant::Instant`.

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

No. I've used the compiled artifact on `wasm32-unknown-unknown` but I'm still trying to figure out how to write UT for that artifact.

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

There is no difference on platforms other than `wasm32-unknown-unknown`

alamb reviewed (commented):
Thanks @waynexia 

I took a look through the Cargo.lock in my datafusion checkout and it appears that adding `instant` would be a new dependency. I am always worried when we potentially increase the dependencies of DataFusion

I, would it be possible to add a `wasm` feature flag to datafusion and only pull in the needed libraries optionally?

Having a wasm feature flag might make it easier to disable other parts of the code that aren't compatible with WASM

For example, we could have something like
```rust
# in datafusion/common/src/wasm.rs

/// DataFusion wrapper around std::instant
/// uses std::time::Instant normally, but uses
/// instant::Instant library when targeting web assembly
struct Instant { 
...
}
```

🤔 


It would also really be nice to get some sort of test / way to run wasm to trigger a bug without changes

I wonder if we could extend https://github.com/apache/arrow-datafusion/tree/main/datafusion/wasmtest 🤔

alamb commented:
Marking as draft as I don't think this PR is waiting on review anymore and I am trying to clear the review queue.

waynexia commented:
Thanks for your review @alamb 👍 

Enclose wasm-specific dependencies and logic with a feature gate looks great to me. I would warp one `Instant` struct for this case.

waynexia commented:
I've extracted all the references into `datafusion_common` and added a target family gateway. Now the implementation of `Instant` will switch automatically when compiling under `wasm32-unknown-unknown`.

I also extended `wasmtest` to include unit tests, which can be run by `wasm-pack test`. Refer to `datafusion/wasmtest/README.md` for detailed commands. Noticed that I haven't added CI for it, because the `--headless` mode has some unknown problem and would timeout in my environment. Once I fix it we can have a CI about wasm by following the link I added in `README.md`.

The lock file of `datafusion-cli` still contains `instant` crate. This seems to be a bug of cargo: https://github.com/rust-lang/cargo/issues/10801. I've tested that it won't be used in actual.

waynexia commented on datafusion/wasmtest/src/lib.rs:
```diff
@@ -61,10 +61,67 @@ pub fn try_datafusion() {
+        let session_context =
+            Arc::new(SessionContext::new_with_config_rt(session_config, rt));
+
+        let statement = DFParser::parse_sql(&sql).unwrap().pop_back().unwrap();
```
```suggestion
        let statement = DFParser::parse_sql(sql).unwrap().pop_back().unwrap();
```

alamb reviewed (approved):
Thanks @waynexia  -- I think this looks good to me. I wonder if you want to put an "exclude" type lint in to prevent people from accidentally using `std::time::Instant`

Similar to how @DDtKey  did this in https://github.com/apache/arrow-datafusion/blob/c439bc73b6a9ba9efa4c8a9b5d2fb6111e660e74/clippy.toml#L1-L4 / https://github.com/apache/arrow-datafusion/pull/9318

DDtKey commented on datafusion/common/src/wasm.rs:
Just one question: 
it's wrapper for `Instant`, but module name is `wasm` while "wasm" it or not depends on target family
A bit confusing to use `wasm::Instant` in code, like it's something specific  (e.g https://github.com/apache/arrow-datafusion/blob/857a2e34299b7ec5be97b40a5019b15d0fc23bd0/datafusion/core/benches/parquet_query_sql.rs#L28)

Shouldn't it be `common::instant::Instant` instead? This way we fully encapsulate wrapper

alamb replied:
> Shouldn't it be `datafusion_common::instant::Instant` instead? This way we fully encapsulate wrapper and can easily extend in the future if we wish


I agree this would be a better home

waynexia replied:
Makes sense 👍 please see [c823e38](https://github.com/apache/arrow-datafusion/pull/9189/commits/c823e3885b09a5df1e5508ab32813e3a7f17ca7a)

DDtKey commented:
>I wonder if you want to put an "exclude" type lint in to prevent people from accidentally using std::time::Instant

Unfortunately, I think it's not possible "as is" for this case, because it's actually just a type alias. However new-type pattern may work (wrap `Instant` into new type instead of type-alias)

Somebody already tried, btw: https://github.com/rust-lang/rust-clippy/issues/10406

waynexia commented:
> > I wonder if you want to put an "exclude" type lint in to prevent people from accidentally using std::time::Instant
> 
> Unfortunately, I think it's not possible "as is" for this case, because it's actually just a type alias. However new-type pattern would work (wrap `Instant` into new type instead of type-alias)
> 
> Somebody already tried, btw: [rust-lang/rust-clippy#10406](https://github.com/rust-lang/rust-clippy/issues/10406)

TIL, thanks for your information. 

I find `disallowed_type` works for our case: https://github.com/apache/arrow-datafusion/pull/9189/commits/edb67fae06bbddb080fae7e6b57b97d5236f8c56, it seems work by matching the import types and won't follow type alias 😉

DDtKey commented:
> I find disallowed_type works for our case: https://github.com/apache/arrow-datafusion/commit/edb67fae06bbddb080fae7e6b57b97d5236f8c56, it seems work by matching the import types and won't follow type alias 😉

That's nice!

DDtKey commented on datafusion/common/src/instant.rs:
```diff
@@ -0,0 +1,28 @@
+// specific language governing permissions and limitations
+// under the License.
+
+//! WASM-compatible `Instant` warpper.
```
Just a typo 🙂 
```suggestion
//! WASM-compatible `Instant` wrapper.
```

alamb commented on clippy.toml:
```diff
@@ -2,3 +2,7 @@ disallowed-methods = [
 ]
+
+disallowed-types = [
+    { path = "std::time::Instant", reason = "Use `datafusion_common::instant::Instant` instead for WASM compatibility" },
```
👍

alamb reviewed (approved):
Thanks @waynexia  and @DDtKey
